hack/remote-debug/README.md

```diff
@@ -0,0 +1,487 @@
+edit/compile iteration.
+
+## Rationale
+
+Traditionally, iterating on edit/compile/debug is slow because it
+involves building a new image, pushing the image, and waiting for the
+roll out to complete. This new environment provides a one-time deploy
+of the debug image. Subsequent iterations only require editing and
+compiling as before. The debugging step just copies a new binary to
+the pod and starts executing immediately, eliminating the need to wait
+for container builds, tagging an image, pushing an image, setting a
+new image on the deployment, and waiting for new pods to roll out.
+
+## Overview
+
+The debug image runs `tini` as the init system. `tini` launches
+`remote-debug-helper`, which performs the following tasks:
+
+1. Discovers all environment variables associated with the
+   `router-default` deployment.
+
+2. Writes a file named `/etc/profile.d/router-default.sh` that contains all of
+   `router-default`'s environment variables, ensuring arbitrary SSH
+   login shells get an identical environment as with `oc rsh`.
+
+3. Starts `sshd` for SSH access into the pod for development and
+   debugging.
+
+This setup allows you to run the `openshift-router` binary without
+specifying command line flags or environment variables, ensuring a
+friction-free invocation that yields an identical environment as a
+non-debug-based Dockerfile.
+
+## Preparatory Steps
+
+Before you begin, there are several preparatory steps you need to
+follow to set up your development environment.
+
+### 1. Obtain the `dlv` Binary
+
+    The `dlv` (Delve debugger) executable needs to be located in the
+    top-level directory as it is copied into the image during the
+    container build process. Use the provided helper script to find and
+    copy the `dlv` binary:
+
+    ```sh
+    $ ./hack/remote-debug/find-and-copy-dlv
+    ```
+
+    This script should work on FHS-compliant installations of GoLand.
+
+### 2. Set Environment Variables for image build and SSH access
+
+    You need to specify several environment variables for the build and
+    deployment processes. These variables can be set in a `.envrc` file in
+    the top-level directory.
+
+    - **REGISTRY**: The container registry where the image will be pushed.
+
+    - **IMAGE**: The name of the debug image.
+
+    - **GITHUB_USER**: When building the custom debug image with SSH support,
+      only public key authentication is supported for SSH login to the pod.
+      The public keys are fetched from
+      `https://github.com/${GITHUB_USER}.keys`, and the Dockerfile requires
+      this environment variable to be set.
+
+    - **DOCKERFILE_DEBUG**: This ensures that the Makefile uses the correct
+      Dockerfile for building the debug image, enabling features necessary
+      for remote debugging.
+
+#### Example `.envrc` file:
+
+```sh
+export REGISTRY=quay.io
+export IMAGE=amcdermo/openshift-router
+export GITHUB_USER=frobware
```

Comment:
As we saw, the github user may not have the ssh key.

Reply:
Done in 55c2b49793b2011a458b81e812be9c6f4f671b33.